### ISSUE TITLE
refining with an evar incorrectly fails to restrict its context (Error: No such section variable or assumption: x.)

### ISSUE THREAD

**Issue by JasonGross** [coq/coq#9937]
#### Description of the problem

```coq
Goal True -> True /\ True.
  intro x; split;
    unshelve (let f := open_constr:(_) in clear x; refine f); exact x.
Defined.
```

#### Coq Version

8.9.0

**Comment by andres-erbsen** [comment]
Attempting to make sense of it...

```coq
Goal True -> True.
  intro x.
  unshelve (
      let e := open_constr:(_) in
      clear x;
      exact e).
  exact x.
  Show Proof.
(* (fun _ : True => x) *)
Qed.
```

**Comment by JasonGross** [comment]
Note that this can also happen with `rename`:
```coq
Goal True -> True /\ True.
  intro x; split;
    unshelve (let f := open_constr:(_) in rename x into x'; refine f); exact x.
Qed. (* Error: No such section variable or assumption: x. *)
```

Here is a workaround:
```coq
Ltac refine_restrict v :=
  let v' := open_constr:(_) in
  let __ := open_constr:(eq_refl : v = v') in
  refine v'.
Goal True -> True /\ True.
  intro x; split;
    unshelve (let f := open_constr:(_) in clear x; refine_restrict f).
  Fail exact x.
```

**Comment by JasonGross** [comment]
Here is perhaps a slightly more general workaround:
```coq
Ltac restrict v :=
  let v' := open_constr:(_) in
  let __ := open_constr:(eq_refl : v = v') in
  idtac.
Goal True -> True /\ True.
  intro x; split;
    unshelve (let f := open_constr:(_) in clear x; restrict f; refine f).
  Fail exact x.
```

**Comment by andres-erbsen** [comment]
Oh, you think expected behavior is for the tactic to fail? In #9937 removing the `rename` made the proof pass.

**Comment by andres-erbsen** [comment]
Is there a reason why errors at Qed are not considered anomalies?

**Comment by SkySkimmer** [comment]
Unsafe tactics like exact_no_check exist

Gaëtan Gilbert

**Comment by JasonGross** [comment]
It is better for the proof to fail early than for it to not fail until Qed.  I don't see how you can get Coq to propagate the renaming inside the evar contexts when it doesn't track all of the evars floating around in Ltac variables and thunks.  If you want to leave the rename, you need to be careful to also issue the rename in all evars that you care about.  (Posing the evar in the context might be enough to get Coq to do this automatically?)

**Comment by andres-erbsen** [comment]
if the proofs are bad, can we get to qed with something that shouldn't be provable? if we can't, are the proofs actually bad?

**Comment by JasonGross** [comment]
There's not enough information left to do that by the time you get to Qed.  What's the bad thing that's unprovable when your proof is `λ x. y` with no `y` in the global environment because you refined with an ever in a context where `y` had been renamed to `x` but the context tracking lost track of that?  And anyway, this should not happen without using `*_no_check`; Coq should be more careful about refining with evars.

**Comment by jadephilipoom** [comment]
Just bumping this thread -- I also ran into this issue and found it problematic, because then I had to go through a large proof trying to figure out where the illegal `instantiate` was. Would be nice to have an earlier error.

**Comment by SkySkimmer** [comment]
Seems to have been fixed by 30aaab9efe02120d286c7f855cc9d855ae11e829 (#18505)
